### PR TITLE
Add locale on users creation

### PR DESCRIPTION
**Pull request by georgestam**
@vemv I have added a `locale` string variable to `users`. In this way, the platform will know whether the user speaks Arabic or English, when the user is created.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
+  LANGUAGES = %w[ar en].freeze
+  
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable and :omniauthable
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :trackable, :validatable
 
   validates :email, email_format: { message: "doesn't look like an email address" }, presence: true
   validates :name, presence: true
+  validates :locale, inclusion: { in: LANGUAGES, allow_nil: false }
```
Comment (vemv) [inline app/models/user.rb]: change this for:

`inclusion: LANGUAGES`

That forbids nil too

File: config/initializers/devise.rb [hunk 2]
```diff
@@ -5,7 +5,7 @@
   config.email_regexp = /\A[^@\s]+@[^@\s]+\z/
   config.expire_all_remember_me_on_sign_out = true
   config.mailer_sender = "rakan@starfishproject.ae"
-  config.password_length = Rails.env.development? ? 1..72 : 10..255
+  config.password_length = Rails.env.development? ? 1..72 : 8..255
```
Comment (vemv) [inline config/initializers/devise.rb]: production passwords should have 10 at least, 8 is too few

File: spec/features/user_registration_spec.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+describe "User registration" do
+  
+  let(:reference){ FactoryGirl.build :user}
+  
+  before {
+    visit root_path
+  }
+  
+  it 'works' do
```
Comment (vemv) [inline spec/features/user_registration_spec.rb]: try to keep `'works'` at a minimum.

Instead:

`it 'results in a User being created'`

File: app/controllers/registrations_controller.rb [hunk 4]
```diff
@@ -3,7 +3,7 @@ class RegistrationsController < Devise::RegistrationsController
   private
 
   def sign_up_params
-    params.require(:user).permit(:name, :email, :password, :password_confirmation)
+    params.require(:user).permit(:name, :email, :password, :password_confirmation).merge(:locale => I18n.locale)
```
Comment (vemv) [inline app/controllers/registrations_controller.rb]: as part of this PR, change `before_action :set_locale` with `around_action :set_locale`, and the following implementation:

```ruby
  def set_locale
    I18n.locale = params[:locale] || I18n.default_locale
    yield
    I18n.locale = I18n.default_locale
  end
```

The point is to not let Arabic users change the global locale for all users.

**Review by vemv — CHANGES_REQUESTED** [review]

File: app/controllers/registrations_controller.rb (hunk 4 above)
Comment (georgestam) [reply app/controllers/registrations_controller.rb]: > The point is to not let Arabic users change the global locale for all users.

What do you mean by 'all users'? how can a user change the 'global locale' for all the users? 

I think that understand how `around_action`works but I have difficulties to see its practical applicability in this (or other) example...

**Comment by georgestam** [comment]
@vemv I have implemented the changes.

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -43,4 +45,13 @@ def destroy_user_if_current_user
     @user = User.new
   end 
   
+  private
+  def extract_locale_from_accept_language_header
+    if language = request.env['HTTP_ACCEPT_LANGUAGE']
```
Comment (georgestam) [inline app/controllers/application_controller.rb]: Rubocop complains about this line, how can I refactor this?`app/controllers/application_controller.rb:50:17: W: Assignment in condition - you probably meant to use ==`
`if language = request.env['HTTP_ACCEPT_LANGUAGE']`

File: spec/features/locales_spec.rb [hunk 6]
```diff
@@ -5,7 +5,7 @@
   }
     
   it 'changes to locale in Arabic when Arabic button is clicked' do
-    expect{ find("#change-to-arabic").click }.to change{ I18n.locale }.from(:en).to(:ar)
+    expect{ find("#change-to-arabic").click }.to change{ page.current_path }.from("/en").to("/ar")
     page_ok
```
Comment (georgestam) [inline spec/features/locales_spec.rb]: I spent a lot of time trying to use the `lang` attribute from the `html` tag without success. So, I implemented this solution instead.

**Review by georgestam — COMMENTED** [review]
@vemv I have added a couple of comments

File: spec/features/locales_spec.rb [hunk 7]
```diff
@@ -5,7 +5,7 @@
   }
     
   it 'changes to locale in Arabic when Arabic button is clicked' do
-    expect{ find("#change-to-arabic").click }.to change{ I18n.locale }.from(:en).to(:ar)
+    expect{ find("#change-to-arabic").click }.to change{ page.current_path }.from("/en").to("/ar")
```
Comment (vemv) [inline spec/features/locales_spec.rb]: Good! but you could additionally add:

`.and change{ find('html')[:lang] }.from('en').to('ar')`

You can find and example of rspec `.and` in kuina.

File: app/controllers/registrations_controller.rb [hunk 8]
```diff
@@ -3,7 +3,8 @@ class RegistrationsController < Devise::RegistrationsController
   private
 
   def sign_up_params
-    params.require(:user).permit(:name, :email, :password, :password_confirmation)
+    password = SecureRandom.hex(10)
```
Comment (vemv) [inline app/controllers/registrations_controller.rb]: not good, much less without a comment

File: app/controllers/application_controller.rb [hunk 9]
```diff
@@ -43,4 +45,13 @@ def destroy_user_if_current_user
     @user = User.new
   end 
   
+  private
+  def extract_locale_from_accept_language_header
```
Comment (vemv) [inline app/controllers/application_controller.rb]: * Only return `locale` if `locale.in?(User::LANGUAGES)`

* Cover this method with a spec. Does it correctly derive `en` from `en`, `en-US`, etc?

* For easier testability, you can extract this method to a new dedicated class like `app/lib/http_accept_lang_parser.rb`. That class would accept a string as an argument, passed from ApplicationController:

```
def extract_locale_from_accept_language_header
  HttpAcceptLangParser.parse(request.env['HTTP_ACCEPT_LANGUAGE'])
end
```

That way in your unit tests you only test HttpAcceptLangParser and don't have to worry about instantiating ApplicationController or manipulating `request`

File: app/models/user.rb [hunk 10]
```diff
@@ -1,11 +1,14 @@
...
+  LANGUAGES = %w[ar en].freeze
+  
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable and :omniauthable
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :trackable, :validatable
 
   validates :email, email_format: { message: "doesn't look like an email address" }, presence: true
   validates :name, presence: true
+  validates :locale, inclusion: { in: LANGUAGES }
```
Comment (vemv) [inline app/models/user.rb]: `inclusion: LANGUAGES`, just like that

**Review by vemv — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb (hunk 5 above)
Comment (vemv) [reply app/controllers/application_controller.rb]:
```
language = request.env['HTTP_ACCEPT_LANGUAGE'].presence
if language
```

File: spec/features/locales_spec.rb (hunk 6 above)
Comment (vemv) [reply spec/features/locales_spec.rb]: Doesn't `find('html')[:lang]` work?

File: app/controllers/registrations_controller.rb (hunk 8 above)
Comment (georgestam) [reply app/controllers/registrations_controller.rb]: I am not creating users to login, I just want to add them to the database (but I need one admin). Before I had hidden fields with the password in the `simple_for,` I think that this approach looks better.

**Comment by georgestam** [comment]
@vemv I have implemented the changes.

File: app/controllers/application_controller.rb [hunk 11]
```diff
@@ -43,4 +45,15 @@ def destroy_user_if_current_user
...
   end 
   
+  private
+  
+  def extract_locale_from_accept_language_header
+    language = request.env['HTTP_ACCEPT_LANGUAGE'].presence
+    if language
+      locale = HttpAcceptLangParser.parse(language)
+      logger.debug "* Locale set using accept_language to '#{locale}'"
+      locale.in?(User::LANGUAGES) ? locale : nil
```
Comment (vemv) [inline app/controllers/application_controller.rb]: move this conditional logic to HttpAcceptLangParser, so all logic (knowledge) is centralized in a single place

File: app/controllers/registrations_controller.rb (hunk 8 above)
Comment (vemv) [reply app/controllers/registrations_controller.rb]: Creating a signup page just for admin is a) overkill, and b) a security risk. Create them in seeds or `rails c`.

You can keep an `edit` page so admins can update their password, but that should be it

Comment (vemv) [reply app/controllers/registrations_controller.rb]: One lesson that one can extract is that workaround code is rarely necessary. Attack causes not symptoms!

File: spec/features/accept_language_spec.rb [hunk 12]
```diff
@@ -0,0 +1,9 @@
+describe "extract locale from client when user has a browser setup in en-US" do
+  
+  let(:accept_language){ "en-US,en;q=0.8,es-419;q=0.6,es;q=0.4,ca;q=0.2,af;q=0.2" }
+    
+  it "parses Accept-Language header into locale string en" do
```
Comment (vemv) [inline spec/features/accept_language_spec.rb]: This is a good start, but remember the golden rule: tests must be **correct** and **complete**.

This is clearly not complete, it only covers a single string. Does it work for `ar` too?

Can you gather a few extra real examples and add them here?

File: spec/features/locales_spec.rb [hunk 13]
```diff
@@ -5,8 +5,16 @@
...
     
   it 'changes to locale in Arabic when Arabic button is clicked' do
-    expect{ find("#change-to-arabic").click }.to change{ I18n.locale }.from(:en).to(:ar)
+    expect{ 
+      find("#change-to-arabic").click 
+    }.to change { 
+      page.current_path 
+    }.from("/en").to("/ar").and change {
+      find('html')[:lang] 
+    }.from("en").to("ar")
```
Comment (vemv) [inline spec/features/locales_spec.rb:14]: Awesome!

File: app/controllers/application_controller.rb (hunk 11 above)
Comment (vemv) [reply app/controllers/application_controller.rb]: Additionally: beware of comparing strings with symbols. What if tomorrow your code looked like this?

`LANGUAGES = %i[ar en].freeze`

(array of symbols)

Such a change is plausible, given that `I18n.locale` always returns a symbol, and as you gain experience you'll use symbols more frequently.

So, a safe comparison would be:

`locale.to_s.in?(User::LANGUAGES.map(&:to_s))`

**Review by vemv — CHANGES_REQUESTED** [review]
Almost there...

File: app/controllers/application_controller.rb [hunk 14]
```diff
@@ -43,4 +45,15 @@ def destroy_user_if_current_user
...
   end 
   
+  private
+  
+  def extract_locale_from_accept_language_header
+    language = request.env['HTTP_ACCEPT_LANGUAGE'].presence
+    if language
+      locale = HttpAcceptLangParser.parse(language)
+      logger.debug "* Locale set using accept_language to '#{locale}'"
+      locale.in?(User::LANGUAGES.map(&:to_sym)) ? locale : nil
```
Comment (georgestam) [inline app/controllers/application_controller.rb]: I noticed that was better to convert it to symbol rather to string, as the `I18n.locale` is returning a symbol (somehow both, symbol and string, seem to work).

File: app/controllers/registrations_controller.rb (hunk 8 above)
Comment (georgestam) [reply app/controllers/registrations_controller.rb]: I did not mean to have a login feature for an admin, what I meant is that I would like to use devise and the users table for both admins and users. Now users do not login, but it is likelty the they will be soon, so I wanted to keep the feature as it is with users/devise.

**Review by georgestam — COMMENTED** [review]
@vemv I have implemented the changes.

File: app/controllers/application_controller.rb (hunk 14 above)
Comment (vemv) [reply app/controllers/application_controller.rb]: The bad part is that `nil.to_sym` raises an error.

It's good do define and return things as symbols. Keep that. But when comparing, always convert both parts to string.

This rule only applies to strings/symbols - not integers or any other type

Comment (vemv) [reply app/controllers/application_controller.rb]: > locale.in?(User::LANGUAGES.map(&:to_sym)) ? locale : nil

remember to move this logic to HttpAcceptLangParser.

File: app/controllers/registrations_controller.rb (hunk 8 above)
Comment (vemv) [reply app/controllers/registrations_controller.rb]: Sounds good, just think of a way of eliminating the need for `SecureRandom.hex(10)`.

**Review by vemv — CHANGES_REQUESTED** [review]

File: app/controllers/registrations_controller.rb (hunk 8 above)
Comment (georgestam) [reply app/controllers/registrations_controller.rb]: Yes, it was an easier way! I added this to `user.rb`: 
```
def password_required?
    super if self.admin
 end

**Review by georgestam — COMMENTED** [review]
@vemv done!

File: app/models/user.rb [hunk 15]
```diff
@@ -1,16 +1,23 @@
...
   validates :name, presence: true
+  validates :locale, inclusion: LANGUAGES
 
   after_create :send_welcome_email
   after_create :subscribe_to_newsletter, if: :production_or_staging?
 
   private
+  
+  def password_required?
+    super if self.admin
```
Comment (vemv) [inline app/models/user.rb:19]: Looks good, but why does the method return `nil` for non-admins?

Couldn't you return `true` for everyone instead?

**Comment by georgestam** [comment]
@vemv I have implemented your change. Is there any reason why do you prefer to show `nil` instead of `false`?

**Comment by vemv** [comment]
You misunderstood my question.

This was the important part:

> Couldn't you return true for everyone instead?

**Pull request merged**